### PR TITLE
fs: allow setting a write buffer for multithread

### PR DESCRIPTION
**Pull request by jorjao81**
#### What is the purpose of this change?

This change adds an option to buffer multithread download parts before writing to disk.

In our testing, we saw that multithreaded downloads from S3 to CIFS (as a mounted filesystem) were much slower with rclone than with the official aws cli. Interestingly, single-threaded downloads were faster than multi-threaded, albeit still far away from what our bandwidth should offer. Downloads to local disk were fine.

The end result looked like this for us:

rclone standard: ~50 MiB/s
rclone with --multi-thread-buffer-size=1M: 400MiB/s+ 

this is close to our max theoretical speed based on bandwidth (5Gib/s)

I did consider just using the already existing buffer-size option, as it does seem to to exactly the same, only for multiple transfers. However I wasn't 100% sure, and with the extra option there should be *no changes* in behaviour unless the user uses the flag, which would not be the case with the buffer-size as it does have a non-zero default AFAIK. Let me know if I should change.

#### Was the change discussed in an issue or in the forum before?

no

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Review by ncw — COMMENTED** [review]
Thank you for doing this - great improvment.

Nice use of Go io primitives. I'd not met `io.NewOffsetWriter` before.

I think the approach looks fine. I think we should probably discuss the options and flags though.

You saw marked improvement with a 1M buffer which is fantastic. That's a small enough number that it could easily be the default.

Did you try different sizes to see what effect that had? Also Windows vs Linux would be very interesting too.

I think since the performance improvements are so large we should probably set the buffer size to non-zero by default, perhaps even to --buffer-size.

I don't think the multithread code is using the read ahead buffer set with --buffer-size so potentially we could - say - divide --buffer-size by number of threads and use that as the buffer size.

Thoughts?

**Comment by jorjao81** [comment]
I just noticed the 1.19 and 1.18 builds failed because of my usage of the OffsetWriter :(

How big of an issue is that? The OffsetWriter was quite nice, as it also allowed me to use the BufferedWriter, so I don't need to do byte arithmetic myself. But if making it go 18 compatible is required for getting the code merged, I can look at it.

**Comment by ncw** [comment]
Your new code looks good.

How did you get on with perf testing with different buffer sizes? Want me to help here?

**Comment by jorjao81** [comment]
Additional test, same configs as the previous:

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `rclone-smb-4-threads-16K` | 44.116 ± 16.421 | 31.497 | 62.682 | 1.38 ± 0.56 |
| `rclone-smb-4-threads-1M` | 37.972 ± 1.404 | 36.351 | 38.808 | 1.19 ± 0.19 |
| `rclone-smb-4-threads-10M` | 34.916 ± 4.469 | 30.027 | 38.792 | 1.09 ± 0.22 |
| `rclone-smb-4-threads-50M` | 38.654 ± 0.909 | 37.873 | 39.651 | 1.21 ± 0.19 |
| `rclone-smb-8-threads-16K` | 37.537 ± 2.010 | 35.223 | 38.850 | 1.18 ± 0.19 |
| `rclone-smb-8-threads-1M` | 37.757 ± 0.177 | 37.606 | 37.952 | 1.18 ± 0.18 |
| `rclone-smb-8-threads-10M` | 37.737 ± 0.953 | 37.074 | 38.829 | 1.18 ± 0.18 |
| `rclone-smb-8-threads-50M` | 39.745 ± 0.599 | 39.079 | 40.238 | 1.25 ± 0.19 |
| `rclone-smb-20-threads-16K` | 31.898 ± 4.844 | 28.963 | 37.490 | 1.00 |
| `rclone-smb-20-threads-1M` | 33.098 ± 4.187 | 28.287 | 35.913 | 1.04 ± 0.21 |
| `rclone-smb-20-threads-10M` | 37.702 ± 0.098 | 37.592 | 37.780 | 1.18 ± 0.18 |
| `rclone-smb-20-threads-50M` | 37.100 ± 1.278 | 35.960 | 38.481 | 1.16 ± 0.18 |
| `rclone-smb-50-threads-16K` | 33.467 ± 4.258 | 28.558 | 36.151 | 1.05 ± 0.21 |
| `rclone-smb-50-threads-1M` | 33.114 ± 4.513 | 29.340 | 38.113 | 1.04 ± 0.21 |
| `rclone-smb-50-threads-10M` | 34.151 ± 4.283 | 29.217 | 36.915 | 1.07 ± 0.21 |
| `rclone-smb-50-threads-50M` | 37.928 ± 2.900 | 34.665 | 40.210 | 1.19 ± 0.20 |

**Comment by jorjao81** [comment]
### Server 
AWS m6in.2xlarge with Windows
gp3 EBS volume
5000 IOPS
500 throughput
(same as before)

### Client (Windoes)
m6in.xlarge

### Test command
    hyperfine -r 3 --show-output -n "rclone-smb-50-thread-{buffer}" --parameter-list buffer 4k,16K,32k,64k,128k,256K,512k,1M,2M,5M,10M 'rclone -vv copyto s3:MYBUCKET/test-10-GiB test --multi-thread-streams 8 --progress --multi-thread-write-buffer-size={buffer} && del /F /Q test ' --export-markdown results.md

### Results

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `rclone-smb-50-thread-4k` | 159.705 ± 4.533 | 154.474 | 162.474 | 6.45 ± 0.44 |
| `rclone-smb-50-thread-16K` | 100.276 ± 0.459 | 99.828 | 100.745 | 4.05 ± 0.25 |
| `rclone-smb-50-thread-32k` | 64.814 ± 2.513 | 62.850 | 67.645 | 2.62 ± 0.19 |
| `rclone-smb-50-thread-64k` | 40.437 ± 1.068 | 39.206 | 41.114 | 1.63 ± 0.11 |
| `rclone-smb-50-thread-128k` | 28.261 ± 1.093 | 27.375 | 29.482 | 1.14 ± 0.08 |
| `rclone-smb-50-thread-256K` | 26.881 ± 4.588 | 21.833 | 30.795 | 1.08 ± 0.20 |
| `rclone-smb-50-thread-512k` | 26.736 ± 1.400 | 25.856 | 28.350 | 1.08 ± 0.09 |
| `rclone-smb-50-thread-1M` | 25.996 ± 1.367 | 24.426 | 26.925 | 1.05 ± 0.09 |
| `rclone-smb-50-thread-2M` | 25.000 ± 1.454 | 23.976 | 26.664 | 1.01 ± 0.09 |
| `rclone-smb-50-thread-5M` | 26.120 ± 1.493 | 24.408 | 27.158 | 1.05 ± 0.09 |
| `rclone-smb-50-thread-10M` | 24.780 ± 1.539 | 23.807 | 26.554 | 1.00 |

**Comment by jorjao81** [comment]
Hi, see the windows tests above, they also seem to benefit quite a bit from a bigger buffer. That said, I think that the most common case by far will be writing to SSDs, am I right?

That being the case, I would go for 16k as the default, and explain that for HDDs and mounted remote filesystems it can pay to test higher numbers.
However, I sorta did change my mind and would to do it per thread though, as I think that will usually be the relevant factor. Do you agree? And in that case, it cannot really use --buffer-size but need our own flag...

**Comment by ncw** [comment]
> Hi, see the windows tests above, they also seem to benefit quite a bit from a bigger buffer.

Yes they do. 128k would seem to be a better default for those.

> That said, I think that the most common case by far will be writing to SSDs, am I right?

Most likely.

> That being the case, I would go for 16k as the default

I think that the feature makes very little difference on Linux - the changes are lost in the noise I think so 128k would be fine there too .

As far as memory use - no-one is going to notice an extra 4*128k = 0.5MB per multi-thread copy.

>  and explain that for HDDs and mounted remote filesystems it can pay to test higher numbers. However, I sorta did change my mind and would to do it per thread though, as I think that will usually be the relevant factor. Do you agree? And in that case, it cannot really use `--buffer-size` but need our own flag...

Yes I think the flag as you have it now is probably exactly what we need.


Here are the results from my Windows VM running on SSD which qualitatively speaking aren't much different from yours though my machine has lots of variance in the runs (its a busy dev machine!).

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `rclone-win-ssd-4-threads-0` | 3.311 ± 0.408 | 2.862 | 3.894 | 1.59 ± 0.20 |
| `rclone-win-ssd-4-threads-16k` | 3.687 ± 0.794 | 2.777 | 4.744 | 1.77 ± 0.38 |
| `rclone-win-ssd-4-threads-32k` | 2.960 ± 0.207 | 2.830 | 3.328 | 1.42 ± 0.10 |
| `rclone-win-ssd-4-threads-64k` | 2.728 ± 0.464 | 2.441 | 3.540 | 1.31 ± 0.22 |
| `rclone-win-ssd-4-threads-128k` | 2.352 ± 0.049 | 2.302 | 2.418 | 1.13 ± 0.03 |
| `rclone-win-ssd-4-threads-256k` | 2.203 ± 0.074 | 2.133 | 2.314 | 1.06 ± 0.04 |
| `rclone-win-ssd-4-threads-512k` | 2.149 ± 0.080 | 2.074 | 2.262 | 1.03 ± 0.04 |
| `rclone-win-ssd-4-threads-1M` | 2.130 ± 0.044 | 2.077 | 2.184 | 1.02 ± 0.03 |
| `rclone-win-ssd-4-threads-2M` | 2.083 ± 0.034 | 2.047 | 2.126 | 1.00 |
| `rclone-win-ssd-4-threads-4M` | 2.116 ± 0.063 | 2.051 | 2.196 | 1.02 ± 0.03 |
| `rclone-win-ssd-4-threads-8M` | 2.565 ± 0.489 | 2.179 | 3.318 | 1.23 ± 0.24 |

So in conclusion, I think this feature is great and the flag you've made already is what we should use.

I'd like to set the default to something > 0. 16k maybe or 128k maybe - I'd be happy with either or something in between - I leave the choice to you :-)

I think that plus a little extra in the help about HDD and/or network file systems and this will be ready to merge :-)

**Comment by jorjao81** [comment]
I went with 128k, as the benefits for Windows seem significant and I don't think there is any real drawback for other cases as you mentioned.

Also update the docs, hope it's clear enough.

Let me know if there is anything else I should do.

Thanks!

**Review by ncw — APPROVED** [review]
Looks great - thank you very much :-)

I'll merge ~~now~~ when the CI goes green!

**Comment by ncw** [comment]
Thank you :-)
